pkg/flow/streaming/flow_sliding_window.go

```diff
@@ -0,0 +1,311 @@
+	}
+	// start processing
+	s.Add(1)
+	go s.receive()
+
+	return nil
+}
+
+func (s *TumblingTimeWindows) purgeWindow(w timeWindow) {
+	s.snapshotsMu.Lock()
+	if snapshot, ok := s.snapshots[w.MaxTimestamp()]; ok {
+		// get and remove the key/value
+		delete(s.snapshots, w.MaxTimestamp())
+		s.snapshotsMu.Unlock()
+		if err := s.acc.Merge(snapshot); err != nil {
+			s.errorHandler(err)
+			return
+		}
+
+		s.out <- flow.NewStreamRecord(s.acc.Snapshot(), w.start)
+		return
+	}
+	s.snapshotsMu.Unlock()
+}
+
+func (s *TumblingTimeWindows) purgeOutdatedWindows() {
+	s.timerMu.Lock()
+	defer s.timerMu.Unlock()
+	for {
+		if lookAhead, ok := s.timerHeap.Peek().(*internalTimer); ok {
+			if lookAhead.triggerTimeMillis <= s.currentWatermark {
+				oldestTimer := heap.Pop(s.timerHeap).(*internalTimer)
+				s.purgeWindow(oldestTimer.w)
+				continue
+			}
+		}
+		return
+	}
+}
+
+func (s *TumblingTimeWindows) receive() {
+	defer s.Done()
+
+	for elem := range s.in {
+		assignedWindows, err := s.AssignWindows(elem.TimestampMillis())
+		if err != nil {
+			s.errorHandler(err)
+			continue
+		}
+		ctx := triggerContext{
+			delegation: s,
+		}
+		for _, w := range assignedWindows {
+			// drop if the window is late
+			if s.isWindowLate(w) {
+				continue
+			}
+			tw := w.(timeWindow)
+			ctx.window = tw
+			// add elem to the bucket
+			s.snapshotsMu.Lock()
+			if oldAggr, ok := s.snapshots[tw.MaxTimestamp()]; ok {
+				oldAggr.Add([]flow.StreamRecord{elem})
+			} else {
+				newAggr := s.aggregationFactory()
+				newAggr.Add([]flow.StreamRecord{elem})
+				s.snapshots[tw.MaxTimestamp()] = newAggr
+			}
+			s.snapshotsMu.Unlock()
+
+			result := ctx.OnElement(elem)
+			if result == FIRE {
+				s.purgeWindow(tw)
+			}
+		}
+
+		// even if the incoming elements do not follow strict order,
+		// the watermark could increase monotonically.
+		if elem.TimestampMillis() > s.currentWatermark {
+			s.currentWatermark = elem.TimestampMillis()
+
+			// TODO: add various strategies to allow lateness items, which come later than the current watermark
+			// Currently, assume the current watermark is t,
+			// then we allow lateness items coming after t-windowSize+slideSize, but before t,
+			// i.e. [t-windowSize+slideSize, t)
+			// NOTE: This call is the normal path to flush outdated windows immediately
+			// without considering lateness.
+			// The purged windows may be accumulated again due to non-zero lateness.
+			s.purgeOutdatedWindows()
+		}
+	}
+	close(s.out)
+}
+
+// isWindowLate checks whether this window is valid
+func (s *TumblingTimeWindows) isWindowLate(w flow.Window) bool {
+	return w.MaxTimestamp()+s.lateness <= s.currentWatermark
+}
+
+func (s *TumblingTimeWindows) Teardown(ctx context.Context) error {
+	s.Wait()
+	return nil
+}
+
+func (s *TumblingTimeWindows) Exec(downstream flow.Inlet) {
+	s.Add(1)
+	go flow.Transmit(&s.ComponentState, downstream, s)
+}
+
+func NewTumblingTimeWindows(size time.Duration) *TumblingTimeWindows {
```

Comment:
An interval will generate several windows based on the slide size. The final result relays on the last window which might contain a subset of the complete list.

The lateness flag will cause an unexpected result. For example, if a single element comes late, the flow creates a new window as the final result. This interval will have a new list with an element instead of the previous one. 

The window should hold all elements in an interval:

* Create a window to represent an interval.
* Receive all elements from this interval, sort them, and leave the top n items based on the strategy.
* Send a snapshot of this window to the sinker on the slide passed. 
* Purge this window when the watermark passes its max time. You can wait a while when the lateness is set to collect some elements come late.

A window can not be created more than once otherwise a new window has to load data flushed lastest.

Reply:
> An interval will generate several windows based on the slide size.

No, it is not correct. Currently, our implementation is Tumbling window. An element can only be assign to a single window instead of multiple ones.

Reply:
```
t.errCh = t.streamingFlow.Window(streaming.NewTumblingTimeWindows(t.slideSize)).
```

But the window's size equals a slide instead of an interval. 

Even the window's size matches an interval. How do you flush based on the slide size of an interval? I found a single size in the current window here.

Reply:
The slide is neccecry. We should flush the snapshot to the disk several times in an interval. Especially for a big interval: one hour, one day, etc.

Reply:
> ```
> t.errCh = t.streamingFlow.Window(streaming.NewTumblingTimeWindows(t.slideSize)).
> ```
> 
> But the window's size equals a slide instead of an interval.

The Sliding size means how often we flush the window in order to persist TopN results. Probably the name is confusing?

> Even the window's size matches an interval. How do you flush based on the slide size of an interval? I found a single size in the current window here.

No. Just as the previous implementation, the flush interval, a.k.a the slideSize, is equal to 40% of the measure interval.

We will assign a proper timestamp based on the measure interval and the flush timestamp for each final TopN result set.

> The slide is neccecry. We should flush the snapshot to the disk several times in an interval. Especially for a big interval: one hour, one day, etc.

The answer for the first sentence is for sure: we still have a "slide". For large interval, we may find a better solution.

Reply:
> A window can not be created more than once otherwise a new window has to load data flushed lastest.

I agree that a window can be only created once. This could produce more precise result. I will check the current impl.

Reply:
all right, I noticed `snapshots` contains a list of aggregation results. A window is just a flush event.

Reply:
This is still an issue.

Reply:
> This is still an issue.

Fixed now. PTAL again.